libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -223,26 +221,52 @@ private CreateMessageOptions ActivityToTwilio(Activity activity)
             return messageOptions;
         }
 
-        /// <summary>
-        /// Converts a query string to a dictionary with key-value pairs.
-        /// </summary>
-        /// <param name="query">The query string to convert.</param>
-        /// <returns>A dictionary with the query values.</returns>
-        private Dictionary<string, string> QueryStringToDictionary(string query)
+        private Activity ReadRequest(HttpRequest httpRequest)
         {
-            var pairs = query.Split('&');
-            var values = new Dictionary<string, string>();
+            var twilioSignature = httpRequest.Headers["x-twilio-signature"];
+            var validationUrl = _options.ValidationUrl ?? (httpRequest.Headers["x-forwarded-proto"][0] ?? httpRequest.Protocol + "://" + httpRequest.Host + httpRequest.Path);
+            var requestValidator = new RequestValidator(_options.AuthToken);
+            var jsonBody = string.Empty;
 
-            foreach (var p in pairs)
+            using (var bodyStream = new StreamReader(httpRequest.Body))
             {
-                var pair = p.Split('=');
-                var key = pair[0];
-                var value = Uri.UnescapeDataString(pair[1]);
+                jsonBody = QueryStringToJson(bodyStream.ReadToEnd());
+            }
 
-                values.Add(key, value);
+            if (string.IsNullOrWhiteSpace(twilioSignature) && requestValidator.Validate(validationUrl, jsonBody, twilioSignature))
```

Comment:
We talked about how the validation by **requestValidator** would fail for **twilioSignature** being null, empty, or whitespace so maybe the left-hand check wouldn't be necessary